drivers/flash/flash_realtek_rts5912.c

```diff
@@ -0,0 +1,826 @@
+}
+
+static inline void spic_usermode(const struct device *dev)
+{
+	const struct flash_rts5912_dev_config *config = dev->config;
+	volatile struct reg_spic_reg *spic_reg = config->regs;
+
+	spic_reg->CTRL0 |= SPIC_CTRL0_USERMD;
+}
+
+static inline void spic_automode(const struct device *dev)
+{
+	const struct flash_rts5912_dev_config *config = dev->config;
+	volatile struct reg_spic_reg *spic_reg = config->regs;
+
+	spic_reg->CTRL0 &= ~SPIC_CTRL0_USERMD;
+}
+
+static void spic_prepare_command(const struct device *dev, const struct qspi_cmd *command,
+				 uint32_t tx_size, uint32_t rx_size, uint8_t write)
+{
+	const struct flash_rts5912_dev_config *config = dev->config;
+	volatile struct reg_spic_reg *spic_reg = config->regs;
+	uint8_t addr_len = user_addr_len[command->address.size];
+
+	spic_flush_fifo(dev);
+
+	/* set SSIENR: deactivate to program this transfer */
+	spic_reg->SSIENR = 0UL;
+
+	/* set CTRLR0: TX mode and channel */
+	spic_reg->CTRL0 &= ~(TMOD(3) | CMD_CH(3) | ADDR_CH(3) | DATA_CH(3));
+	spic_reg->CTRL0 |= TMOD(write == 0x01 ? 0x00UL : 0x03UL) |
+			   ADDR_CH(command->address.bus_width) | DATA_CH(command->data.bus_width);
+
+	/* set USER_LENGTH */
+	spic_reg->USERLENGTH = USER_CMD_LENGTH(1) |
+			       USER_ADDR_LENGTH(command->address.disabled ? 0 : addr_len) |
+			       USER_RD_DUMMY_LENGTH(command->dummy_count * spic_reg->BAUDR * 2);
+
+	/* Write command */
+	if (!command->instruction.disabled) {
+		spic_reg->DR.BYTE = command->instruction.value;
+	}
+
+	/* Write address */
+	if (!command->address.disabled) {
+		for (int i = 0; i < addr_len; i++) {
+			spic_reg->DR.BYTE =
+				(uint8_t)(command->address.value >> (8 * (addr_len - i - 1)));
+		}
+	}
+
+	/* Set TX_NDF: frame number of Tx data */
+	spic_reg->TXNDF = TX_NDF(tx_size);
+
+	/* Set RX_NDF: frame number of receiving data. */
+	spic_reg->RXNDF = RX_NDF(rx_size);
+}
+
+static void spic_transmit_data(const struct device *dev, const void *data, uint32_t *length)
+{
+	const struct flash_rts5912_dev_config *config = dev->config;
+	volatile struct reg_spic_reg *spic_reg = config->regs;
+
+	uint32_t len = *length;
+
+	/* set SSIENR to start the transfer */
+	spic_reg->SSIENR = SPIC_SSIENR_SPICEN;
+
+	/* write the remaining data into fifo */
+	for (int i = 0; i < len;) {
+		if (spic_reg->SR & SPIC_SR_TFNF) {
+			spic_reg->DR.BYTE = ((const uint8_t *)data)[i];
+			i++;
+		}
+	}
+}
+
+static void spic_receive_data(const struct device *dev, void *data, uint32_t *length)
+{
+	const struct flash_rts5912_dev_config *config = dev->config;
+	volatile struct reg_spic_reg *spic_reg = config->regs;
+
+	uint32_t i, cnt, rx_num, fifo, len;
+	uint8_t *rx_data = data;
+
+	len = *length;
+	rx_data = data;
+
+	/* set SSIENR to start the transfer */
+	spic_reg->SSIENR = SPIC_SSIENR_SPICEN;
+
+	rx_num = 0;
+	while (rx_num < len) {
+		cnt = spic_reg->RXFLR;
+
+		for (i = 0; i < (cnt / 4); i++) {
+			fifo = spic_reg->DR.WORD;
+			memcpy((void *)(rx_data + rx_num), (void *)&fifo, 4);
```

Comment:
@JhanBoChao-Realtek You may want to look into the buffer overflow issue reported by sonarcloud
https://sonarcloud.io/project/issues?severities=BLOCKER%2CCRITICAL&sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&types=BUG&id=zephyrproject-rtos_zephyr&open=AZbTWQofVYyps_414stQ

thanks!